docs/0001-teleport-job-worker.md

```diff
@@ -0,0 +1,185 @@
+---
+authors: Brodie Kurczynski (brodie@kurczynski.org)
+state: draft
+---
+
+# RFD 1 - Teleport Job Worker
+
+## What
+
+A service that runs commands submitted by multiple users. These commands may run concurrently and will provide the users
+with the ability to limit resource (i.e. CPU, memory, and disk IO) usage. The service will consist of three pieces:
+
+* A reusable library implementing the functionality for working with jobs
+* An API server that wraps the functionality of the library
+* A command line interface (CLI) that communicates with the API server
+
+### Required Approvers
+
+* Engineering: (@sclevine && @gavinfrazar) || (@sclevine && @bernardjkim) || (@sclevine && @jimbishopp) || (
+  @gavinfrazar && @bernardjkim) || (@gavinfrazar @jimbishopp) || (@bernardjkim @jimbishopp)
+
+### Security
+
+Requests made to the service will use mTLS for authentication. Specifically:
+
+* TLS using the modern version 1.3
+* The cipher suite will need no configuration, it will be handled by `crypto/tls` (this was updated
+  in [Go 1.17](https://go.dev/doc/go1.17#crypto/tls))
+* X.509 public/private key pairs will be generated using RSA-4096
+
+For authorization, the service will use simple tokens to ensure that jobs can only be accessed and modified by the user
+who created them. This will be done by adding `grpc.UnaryInterceptor` to the API server to check if the user has
+authorization to interact with the resource. For the purposes of this challenge, this type of authorization will be
+fine. However, proper OAuth would need to be implemented for use in production.
+
+### UX
+
+The CLI used in the service will be a simple application that implements the shared library and expose all the
+functionality of the [API](#api). This is what they will look like in the CLI:
+
+* **Start**
+  ```shell
+  job-worker -token "some-token" start -cpu 10 -memory 1000 -io 400 -cmd "/usr/bin/sleep" -args "60"
+  ```
+  The auto-generated ID for the job will be returned if the job is started successfully. The resource arguments use the
```

Comment:
can you elaborate on how the ID is generated and why it's generated that way?

Reply:
Of course. Done.